(feat) propose new release v5.0.0

medbenmakhlouf opened the pull request:
I had the motivation to update the package "highcharts-angular", including the demo app, to the latest version of Angular to take advantage of new reactivity features, now stable as of version 19.

This update will include:

- [X] Replacing lifecycle hooks with Signals for better state management.

- [X] Introducing Input and Output Signals for component communication.

- [X] Utilizing computed and LinkedSignal for derived state.

- [X] Applying effect to handle side effects efficiently.

By adopting these changes, we can ensure our components are more reactive, performant, and easier to maintain in the long term, with improved test coverage to support the updates.

PowerKiKi commented:
That looks like quite an improvement. What aboud getting rid of the module once and for all, since we are breaking the lib anyway ?

Also, since we are doing a major version, I would strongly suggest to incorporate [those suggestions](https://github.com/highcharts/highcharts-angular/issues/163#issuecomment-2146822290) in one way or another. The purpose of that is to make this lib effortlessly lazy load Highchart by default, and to be able to load extra optional features easily in a typed manner. All of this becomes natural if we replace the input for Highcharts by a `provideHighcharts()` function. What do you think ?

medbenmakhlouf commented:
@PowerKiKi, with my latest commits, I was able to move the code from the Highcharts Component to a directive and use that directive as host directives. Now developers can have two different ways to develop their charts.

On the other hand, I saw your comment about lazy-loading the library, and here I tried to come up with an approach using the preferred syntax we see these days in the Angular ecosystem.

I have developed the provideHighChartsModuleFactory() to help lazy-load Highcharts with modules like Gantt, Stock, Map, etc., on demand.

take a look at the demo example and let me know ;)

PowerKiKi commented:
> have two different ways to develop their charts

Can you elaborate on what is the advantage of having a directive ? It seems to me that a Highcharts charts fits especially well the concept of component, and it feels rather awkward as a directive because it replace the entire content of the element where the directive is applied. What real world use-case could be solved with a directive that cannot be solved with the existing component ?

medbenmakhlouf commented:
@PowerKiKi, Let me elaborate on why directives can be advantageous, even for use cases like Highcharts, where a component might seem more intuitive at first glance.

- Directives in Angular are typically used to enhance behavior or manipulate DOM elements dynamically without creating an entirely new encapsulated component. It is also great for situations where you want to add functionality to an existing element or component without replacing it entirely. 

- Directives don’t come with the overhead of template rendering or encapsulation like a component does. If your goal is solely to bind a chart to an element and manage lifecycle hooks efficiently (e.g., initializing, updating, and destroying the chart), a directive is more straightforward.

- With a directive, you can directly apply the charting logic to any container (like a `<div>`), preserving the parent’s structure and style. This flexibility might feel more natural than wrapping everything in a component.

``` html
<div highcharts-charts [options]="chart1Options"></div>  
<div highcharts-charts [options]="chart2Options"></div>  
```

PowerKiKi commented:
I tested your PR at commit f610e0dc02e975724033973708a50c63b6340c68 (because more recent commits seemed to have broken something and even more recent commits were not available yet). I used this code:

```ts
import {Component} from '@angular/core';
import {HighchartsChartDirective} from '../../highcharts-angular/src/lib/highcharts-chart.directive';
import {HighchartsChartComponent} from '../../highcharts-angular/src/lib/highcharts-chart.component';
import type HighchartsESM from 'highcharts/es-modules/masters/highcharts.src';

@Component({
  selector: 'app-root',
  template: `
    <highcharts-chart [options]="options" style="width: 50%; height: 200px; display: block;">This content will never be visible</highcharts-chart>

    <div highcharts-chart [options]="options" style="width: 50%; height: 200px; display: block;">This content will never be visible</div>
  `,
  imports: [HighchartsChartComponent, HighchartsChartDirective],
})

export class AppComponent {
  public options: HighchartsESM.Options = {
    title: {
      text: 'Highcharts chart'
    },
    series: [
      {
        type: 'line',
        data: [11, 2, 3],
        zones: [{
          value: 7.2,
          dashStyle: 'Dot',
          color: 'red'
        }]
      },
      {
        type: 'line',
        data: [5, 6, 7]
      }
    ]
  };
}
```

# Observations

> add functionality to an existing element or component without replacing it entirely

`HighchartsChartComponent` and `HighchartsChartDirective` both replace the entire element. The resulting DOM is nearly identical. Most importantly the content within the element is never projected anywhere, so disappear entirely in both cases.

![image](https://github.com/user-attachments/assets/3e6618d3-2447-4c11-a24d-9c4242b7313d)

> overhead of template rendering or encapsulation like a component does

`HighchartsChartComponent` has no template at all, even in [your most recent commit](https://github.com/highcharts/highcharts-angular/pull/392/files#diff-4d7cf992751846e7e0040dd1c85d509f6d1400443f49c5ea431dd69469bc80a8R6). So there is no overhead neither for template rendering, neither for encapsulation, since there is nothing to encapsulate.

> preserving the parent’s structure and style

My test code shows that style might be applied to both the component and directive version, in the exact same way, with a pixel-perfect identical result.

# Conclusion

All three arguments seems to be moot, or I did not test them the way I should have. If so, please share a example of code using the directive that cannot be replicated with the component.

As it stands, I don't think we should have two different ways to do exactly the same things. This would only increase maintenance burden when people start asking, "Which one should I use ?", like myself.

My 2 cents is we should get rid of the directive. @karolkolodziej, what do you think ?

PowerKiKi commented:
Also the introduction of HighchartsChartService in this PR seems like scope creep to me. And it makes the PR harder to review. I think we should take things step by step in multiple PRs. First introduce signals with ChangeDetectionStrategy.OnPush, then drop support for non-ESM, then solve the lazy loading aspect with providers, and then only add more things on top of it.

medbenmakhlouf commented:
Suppose you have a table where you want to embed Highcharts as background elements (sparklines) within specific table cells. Using a directive, you can dynamically enhance those cells without altering the table structure or encapsulating content into a separate template.


``` html

<table>
  <thead>
    <tr>
      <th>Product</th>
      <th>Sales (Sparkline)</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Product A</td>
      <td highcharts-chart [options]="{ series: [{ data: [5, 10, 15, 10] }] }"></td>
    </tr>
    <tr>
      <td>Product B</td>
      <td highcharts-chart [options]="{ series: [{ data: [20, 30, 25, 35] }] }"></td>
    </tr>
  </tbody>
</table>

```


**Why This Is Difficult to Achieve with a Component**
Preservation of DOM Context

A component would encapsulate the chart, requiring you to replace the <td> entirely with the <highchart-chart> tag. This breaks the table layout or requires complex styling to match the behavior.
With a directive, the chart is applied directly to the <td> without altering the structure.
Behavior Augmentation

Directives are designed to enhance existing elements. Here, highchart-chart directly applies chart behavior to an element that already exists within the DOM.
A component encapsulates behavior and content, which can lead to unwanted nesting in situations like this.
Reusability in Mixed Content

If the <td> contains other child elements (e.g., text or additional icons), a directive can enhance the element without taking over its entire content. For instance:

``` html
<td highchart-chart [options]="chartOptions">
  <span>Custom Label</span>
</td>

```
This level of flexibility is hard to achieve with a component because it replaces or encapsulates the content inside its template.

**Conclusion**

I found that the directive is the best fit here, while the component is still nice to have for building charts. We should keep the component to avoid breaking existing code for people already using the package.

As you can see, I am using this directive as a HostDirective for the HighChartComponent, allowing us to focus on maintaining only the directive moving forward instead of both.

Finally, I am happy to split this PR into smaller ones to make the review and merging process easier.

I will keep pushing code to provide a complete overview of where we’re heading with these changes so that we can align them with different releases.

medbenmakhlouf commented:
@PowerKiKi @karolkolodziej , With my latest commits, which include adding some tests, I have finished what can be done for this release. It’s true that I made significant changes to the library, but I believe these are long-term improvements for future versions of Angular. You can check out my branch, test the demo app, and feel free to provide any feedback. Again, I have marked this PR as a draft since I made several commits, and it might be tough to review. I am happy to split this PR into smaller ones and am also planning to include Angular ESLint to format the code. Thank you again, and I hope this can be a valuable contribution to the community and beneficial for everyone.

karolkolodziej commented:
Hi @medbenmakhlouf and @PowerKiKi! 

Huge thanks for this PR and the whole discussion!
I'm sorry, but I was occupied with something else lately and didn't have time to fully check your changes! 

**At first glance, I believe it will bring enormous value long term!** 

I'll set up some time and do a detailed test and review this week so we can publish it as soon as possible 😉

medbenmakhlouf commented:
Thanks @karolkolodziej for your response! look forward to your feedbacks!

karolkolodziej commented on highcharts-angular/src/lib/highcharts-chart.directive.ts:
```diff
@@ -0,0 +1,127 @@
+  Highcharts = input<Chart['highcharts']>();
+
+  /**
+   * Type of the chart constructor.
```
Add what choice do we have.

medbenmakhlouf replied:
@karolkolodziej I have added ChartConstructorType to provide all constructorTypes, please let me know if I am missing some.

karolkolodziej commented on highcharts-angular/src/lib/highcharts-chart.directive.ts:
```diff
@@ -0,0 +1,127 @@
+  /**
+   * Whether to update the chart one-to-one.
+   */
+  oneToOne = input<boolean>();
```
Could you please add this API doc here: https://api.highcharts.com/class-reference/Highcharts.Chart#update
With sth like:  Check how `update` works in Highcahrt

medbenmakhlouf replied:
@karolkolodziej , how about now ?

karolkolodziej commented on highcharts-angular/src/lib/highcharts-chart.directive.ts:
```diff
@@ -0,0 +1,127 @@
+  options = input<Chart['options']>();
+
+  /**
+   * Whether to rerender the chart.
```
Incorrect description.

medbenmakhlouf replied:
@karolkolodziej done

karolkolodziej commented on highcharts-angular/src/lib/types.ts:
```diff
@@ -0,0 +1,21 @@
+
+export interface Chart {
+  options: Highcharts.Options | HighchartsESM.Options,
+  update: boolean,
```
```suggestion
  update?: boolean,
```

medbenmakhlouf replied:
done

karolkolodziej reviewed (commented):
Tremendous work! Thanks for that 🎉 

A couple of comments: 
1. I'm fine with moving from the component to the directive as long as we have a proper documentation. It took me a while to figure our what is going on because as it was mentioned the component looks and feel more intuitive and natural but I see the benefits of Directives so let's explore it.
2. Could you please move the old documentation to a separate file and start working on a new one in the readme?
3. Could you please check (and if easy add an example) how it will work with SSR?
4. Are you sure we should remove the `runOutsideAngular` flag? 
5. Moving on could we please use Highcharts in v12.x.x? There were some important changes that might break the `providePartialHighChart`

medbenmakhlouf commented on highcharts-angular/package.json:
```diff
@@ -1,7 +1,7 @@
   "name": "highcharts-angular",
   "description": "Highcharts component for Angular.",
-  "version": "4.0.1",
+  "version": "5.0.0",
```
done

medbenmakhlouf commented:
@karolkolodziej  thank you for your support :) , here are my answers to your comments 
> Tremendous work! Thanks for that 🎉
> 
> A couple of comments:
> 
> 1. I'm fine with moving from the component to the directive as long as we have a proper documentation. It took me a while to figure our what is going on because as it was mentioned the component looks and feel more intuitive and natural but I see the benefits of Directives so let's explore it.

I’ve added documentation to clarify the change from component to directive, and I made sure we cover all the important aspects. Let me know if there's anything more you’d like to see or if something is still unclear.

> 2. Could you please move the old documentation to a separate file and start working on a new one in the readme?

I’ve moved the old documentation to a separate file README_V4 and started updating the README. I’ve tried to cover everything, but please let me know if I missed something or if there’s anything else you’d like to add

> 3. Could you please check (and if easy add an example) how it will work with SSR?

I’ve already set up a demo app with SSR, and I’ve made sure all our local showcases work fine. You can try it out with the command: `npm run start:ssr`. Let me know if it works well on your end.

> 4. Are you sure we should remove the `runOutsideAngular` flag?

I’ve changed the component to use the OnPush strategy and everything is now based on Signals, so it should be safe to remove the `runOutsideAngular` flag. see [this](https://angular.dev/guide/signals#reading-signals-in-onpush-components) 

> 5. Moving on could we please use Highcharts in v12.x.x? There were some important changes that might break the `providePartialHighChart`

I’ve tried to upgrade to v12.x.x, but I ran into some issues with webpack when trying to apply the module export. It seems to be taking longer than expected. If you can help me troubleshoot the issue, that would be great. Otherwise, we could aim for a minor version update later that supports v12.0.0. What do you think?

PowerKiKi commented:
> This breaks the table layout or requires complex styling to match the behavior.

Thank you for taking the time to explain this. I agree with that now, as is demonstrated by your example

> directive can enhance the element without taking over its entire content. For instance:
> 
> ```
> <td highchart-chart [options]="chartOptions">
>   <span>Custom Label</span>
> </td>
> ```

I cannot agree with this one though. `Custom Label` will never appear anywhere. This directive will always take over its entire content, which is, I think, contrary to what is usually done in Angular ecosystem. It really does not feel natural, but as you demonstrated a practical use, I can live with that :+1:.

# Build broken

Running `rm -rf yarn.lock && yarn && yarn start` on commit 7c8074093b015367d9c9f65b7189947c585599e1 will fail compilation:

```
Application bundle generation failed. [6.086 seconds]

✘ [ERROR] TS2322: Type 'Promise<typeof import("/sites/3dparty/highcharts-angular/node_modules/highcharts-custom-events/js/customEvents")>' is not assignable to type 'Promise<ModuleFactory>'.
  Type 'typeof import("/sites/3dparty/highcharts-angular/node_modules/highcharts-custom-events/js/customEvents")' is not assignable to type 'ModuleFactory'.
    Types of property 'Highcharts' are incompatible.
      Type 'typeof import("/sites/3dparty/highcharts-angular/node_modules/highcharts-custom-events/node_modules/highcharts/highcharts")' is not assignable to type 'typeof import("/sites/3dparty/highcharts-angular/node_modules/highcharts/highcharts") | typeof import("/sites/3dparty/highcharts-angular/node_modules/highcharts/highcharts.src")'.
        Type 'typeof import("/sites/3dparty/highcharts-angular/node_modules/highcharts-custom-events/node_modules/highcharts/highcharts")' is missing the following properties from type 'typeof import("/sites/3dparty/highcharts-angular/node_modules/highcharts/highcharts.src")': each, grep, inArray, keys, and 4 more. [plugin angular-compiler]

    src/app/line-chart/line-chart.component.ts:10:56:
      10 │ ...ghChart({ modules: () => [import('highcharts-custom-events')] })],
         ╵                              ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

I removed all reference to that package to be able to test the PR.

# Not lazy loaded

`provideHighChartInstance` takes a promise as parameter, and it is an environnement provider. Meaning the provider _must_ be global to the application and highcharts will be loaded as soon as the application initialize. This make lazy loading impossible.

Instead, I suggest we mirror what is done for modules and declare the token as a factory, like below, and adapt the existing code:

```diff
diff --git highcharts-angular/src/lib/types.ts highcharts-angular/src/lib/types.ts
index 2fc8597..2a06e39 100644
--- highcharts-angular/src/lib/types.ts
+++ highcharts-angular/src/lib/types.ts
@@ -3,6 +3,7 @@ import type HighchartsESM from 'highcharts/es-modules/masters/highcharts.src';
 
 export type ChartConstructorType = 'chart' | 'ganttChart' | 'stockChart' | 'mapChart';
 
+export type InstanceFactoryFunction = () => Promise<Chart['highcharts']>;
 export type ModuleFactoryFunction = () => Promise<ModuleFactory>[];
 
 export interface ModuleFactory {
diff --git highcharts-angular/src/lib/highcharts-chart.token.ts highcharts-angular/src/lib/highcharts-chart.token.ts
index 0c69eb7..fb0d2d1 100644
--- highcharts-angular/src/lib/highcharts-chart.token.ts
+++ highcharts-angular/src/lib/highcharts-chart.token.ts
@@ -1,7 +1,7 @@
 import { InjectionToken } from '@angular/core';
-import { Chart, ModuleFactoryFunction, PartialHighchartsConfig } from './types'
+import { Chart, InstanceFactoryFunction, ModuleFactoryFunction, PartialHighchartsConfig } from './types'
 
-export const HIGHCHARTS_LOADER = new InjectionToken<Promise<Chart['highcharts']>>('HIGHCHARTS_LOADER');
+export const HIGHCHARTS_LOADER = new InjectionToken<InstanceFactoryFunction>('HIGHCHARTS_LOADER');
 export const HIGHCHARTS_ROOT_MODULES = new InjectionToken<ModuleFactoryFunction>('HIGHCHARTS_ROOT_MODULES');
 export const HIGHCHARTS_OPTIONS = new InjectionToken<Chart['options']>('HIGHCHARTS_OPTIONS');
 export const HIGHCHARTS_CONFIG = new InjectionToken<PartialHighchartsConfig>('HIGHCHARTS_CONFIG');
```

# Drop CommonJS

ESM is the only correct way of consuming JS library in Angular ecosystem, and thus support for CJS is not welcome anymore. I suggest the following:

```diff
diff --git highcharts-angular/src/lib/types.ts highcharts-angular/src/lib/types.ts
index 2fc8597..b9c46fa 100644
--- highcharts-angular/src/lib/types.ts
+++ highcharts-angular/src/lib/types.ts
@@ -1,4 +1,3 @@
-import type Highcharts from 'highcharts';
 import type HighchartsESM from 'highcharts/es-modules/masters/highcharts.src';
 
 export type ChartConstructorType = 'chart' | 'ganttChart' | 'stockChart' | 'mapChart';
@@ -11,9 +10,9 @@ export interface ModuleFactory {
 }
 
 export interface Chart {
-  options: Highcharts.Options | HighchartsESM.Options,
+  options: HighchartsESM.Options,
   update?: boolean,
-  highcharts?: typeof Highcharts | typeof HighchartsESM
+  highcharts?: typeof HighchartsESM
   constructorChart?: Function;
```

# Drop HighchartsChartModule

I don't see a reason to keep HighchartsChartModule in the next major version. What we _may_ do is deprecated it in a patch version before the major version. But the major version really should drop it entirely. Is has outlived its purpose.

# Minimal, explicit, public API surface

The public API should be kept at a minimum to reduce maintenance and cost and not confuse end-users with things they are not supposed to use directly. Specifically no tokens should be public and internal providers should stay internal. I suggest this:

```diff
diff --git highcharts-angular/src/public_api.ts highcharts-angular/src/public_api.ts
index c4f9103..e344c84 100644
--- highcharts-angular/src/public_api.ts
+++ highcharts-angular/src/public_api.ts
@@ -2,10 +2,14 @@
  * Public API Surface of highcharts-angular
  */
 
-export * from './lib/highcharts-chart.module';
-export * from './lib/highcharts-chart.component';
-export * from './lib/highcharts-chart.directive';
-export * from './lib/highcharts-chart.token';
-export * from './lib/highcharts-chart.provider';
-export * from './lib/highcharts-chart.service';
-export * from './lib/types';
+export {
+  HighchartsChartComponent
+} from './lib/highcharts-chart.component';
+export {
+  HighchartsChartDirective
+} from './lib/highcharts-chart.directive';
+export {provideHighCharts, providePartialHighChart} from './lib/highcharts-chart.provider';
+export {
+  HighchartsConfig,
+  PartialHighchartsConfig,
+} from './lib/types';
```

PowerKiKi commented on highcharts-angular/src/lib/highcharts-chart.directive.ts:
```diff
@@ -0,0 +1,139 @@
+  private highCharts = toSignal(this.highchartsChartService.loaderChanges$);
+
+  private constructorChart = computed<Function>(() => {
+    const constructorType = untracked(this.constructorType);
```
If I understand this correctly, it means that changing `[constructorType]` after the very first initialization will have no effect at all. Is this intentional ? should we be able to change the constructor type while during the component life cycle ? :thinking: 

If it can indeed never be changed later on, then I think it should be documented as such. WDYT ?

medbenmakhlouf replied:
@PowerKiKi , I kept the same behavior as before. If you check the code in [version 4](https://github.com/highcharts/highcharts-angular/blob/ab095ccaa07e5014a1e667506203bcb0c09f4d6a/highcharts-angular/src/lib/highcharts-chart.component.ts#L29), the ngOnChanges method is just tracking the options and update inputs. That’s why I maintained the same behavior. However, I am open to modifying this to track changes whenever this input changes.

medbenmakhlouf commented on highcharts-angular/src/lib/highcharts-chart.directive.ts:
```diff
@@ -0,0 +1,139 @@
+   * Highcharts library or Highcharts ESM module.
+   * @deprecated
+   */
+  Highcharts = input<Chart['highcharts']>();
```
Done, I have added the new section in the readme to indicate the breaking change

medbenmakhlouf commented on highcharts-angular/src/lib/highcharts-chart.directive.ts:
```diff
@@ -0,0 +1,139 @@
+   * Whether to run the chart outside Angular.
+   * @deprecated
+   */
+  runOutsideAngular = input<boolean>();
```
idem

medbenmakhlouf commented on highcharts-angular/src/lib/highcharts-chart.directive.ts:
```diff
@@ -0,0 +1,139 @@
+  /**
+   * Options for the Highcharts chart.
+   */
+  options = input<Chart['options']>();
```
Done

medbenmakhlouf commented on highcharts-angular/src/lib/highcharts-chart.directive.ts:
```diff
@@ -0,0 +1,139 @@
+   * Items are added/removed as needed. Series with `id`s are matched by `id`;
+   * unmatched items are removed. Omitted `series` leaves existing ones unchanged.
+   */
+  oneToOne = input<boolean>();
```
Done

medbenmakhlouf commented on highcharts-angular/src/lib/highcharts-chart.directive.ts:
```diff
@@ -0,0 +1,139 @@
+    callbackFunction: Highcharts.ChartCallbackFunction
+  ): Highcharts.Chart | null {
+    if (chart) {
+      chart.update(source.options, true, oneToOne || false);
```
Done

medbenmakhlouf commented:
@PowerKiKi, thank you for the great feedback and for giving it a try! I have addressed most of your points and decided to include InstanceFactoryFunction as the module factory. Regarding the build issue, I couldn’t reproduce it on my end. Locally, I am using NPM instead of Yarn, so I’m not sure what I might be missing.

all these commands : 

for the library
`npm start`
`npm run build`

for the demo apps
`npm run build:app`
`npm run build:ssr`


cc @karolkolodziej, I’d love to know if you had a chance to look at my latest commit. Thank you both for the excellent feedback!

PowerKiKi commented:
> using NPM instead of Yarn

Switched to NPM and do not see compile error anymore :+1: 

I created a PR for your PR: https://github.com/medbenmakhlouf/highcharts-angular/pull/1

PowerKiKi commented:
@karolkolodziej, have you had the chance to review https://github.com/medbenmakhlouf/highcharts-angular/pull/1 and https://github.com/highcharts/highcharts/pull/22341 which both are prerequisites before reviewing this PR ?
